src/uproot/interpretation/objects.py

```diff
@@ -45,14 +46,21 @@ class AsObjects(uproot.interpretation.Interpretation):
     :ref:`uproot.interpretation.objects.AsObjects.simplify` attempts to
     replace this interpretation with a faster-to-read equivalent, but not all
     data types can be simplified.
+
+    # TODO: known_forth can define forth code and forms for special cases that will be picked up here as well
     """
 
     def __init__(self, model, branch=None):
         self._model = model
         self._branch = branch
-        self._form = None
         self._forth = True
-        self._complete_forth_code = None
+        self._form = None
+        known_forth = known_forth_of(self._model)
+        if known_forth is not None:
+            self._complete_forth_code = known_forth.forth_code
+            self._form = known_forth.awkward_form
+        else:
+            self._complete_forth_code = None
```

Comment:
I think this answers your question:

```suggestion
        known_forth = known_forth_of(self._model)
        if known_forth is not None:
            self._complete_forth_code = known_forth.forth_code
            self._form = known_forth.awkward_form
        else:
            self._complete_forth_code = None
            self._form = None
```

Either the `self._form` and `self._complete_forth_code` are both known or not, and this arrangement of the if-statement makes that more clear.

The part that was confusing me about your question is that you were talking about something _returning_ something. These attributes are being assigned in-place (because it's the constructor).

Reply:
With *returning* i was referring to `AsObjects.awkward_form` returning `self._form` which is something i newly introduced. Before the change, `awkward_form(...)` would always consult `self._model.awkward_form(...)` where now it returns `self._form` that has been provided either from generated or known_forth. The main visible difference is that these forms now also carry the `form_key` attributes for the forth code.